### PR TITLE
tpm2: make it possible to serialize / unserialze PCRProtectionProfile

### PR DESCRIPTION
**Pull request by chrisccoulson**
This makes it possible for PCRProtectionProfile to be serialized to
and unserialized from the TPM wire format, using
github.com/canonical/go-tpm2/mu. The profile is serialized in a fairly
space efficient binary format, where digests in the profile are
de-duplicated.

This might be useful in the future in scenarios where it is desirable
to retain a profile, eg, for debugging purposes.

### CODE REVIEW

File: tpm2/pcr_profile.go [hunk 1]
```diff
@@ -728,6 +735,279 @@ func (p *PCRProtectionProfile) String() string {
...
+}
+
+func (c *pcrProtectionProfileSerializer) extendPCR(alg tpm2.HashAlgorithmId, pcr int, value tpm2.Digest) {
+	c.instrs = append(c.instrs, &savedPCRProtectionProfileInstr{
+		Type: extendPCR,
+		Data: &savedPCRProtectionProfileInstrData{
+			ExtendPCR: &savedPCRProtectionProfilePCREventInstrData{
+				Alg:          alg,
+				PCRAndDigest: newSavedPCRProtectionProfilePCRAndDigest(uint16(pcr), c.digestIndex(value))}}})
+}
```
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply tpm2/pcr_profile.go:888]: Ah, thanks - I'll add some tests for these.

File: tpm2/pcr_profile.go [hunk 2]
```diff
@@ -728,6 +735,279 @@ func (p *PCRProtectionProfile) String() string {
...
+// index and digest index as a uint32 in a saved PCR profile, with the
+// lower 11 bits used to store the PCR index and the upper 22 bits used
+// to store a digest index, which references a digest in a list.
+type savedPCRProtectionProfilePCRAndDigest uint32
+
+func (i savedPCRProtectionProfilePCRAndDigest) PCR() uint16 {
+	return uint16(i) & savedPCRMask
+}
+
+func (i savedPCRProtectionProfilePCRAndDigest) Digest() uint32 {
```
Comment (pedronis) [inline tpm2/pcr_profile.go:755]: shouldn't this and related entities be called DigestIndex if it's an index?

File: tpm2/pcr_profile.go [hunk 3]
```diff
@@ -728,6 +735,279 @@ func (p *PCRProtectionProfile) String() string {
...
+
+// savedPCRProtectionProfileInstrData represents the data associated with a
+// single instruction in a saved PCR profile.
+type savedPCRProtectionProfileInstrData struct {
+	AddPCRValue        *savedPCRProtectionProfilePCREventInstrData
+	AddPCRValueFromTPM *savedPCRProtectionProfileAddPCRValueFromTPMInstrData
+	ExtendPCR          *savedPCRProtectionProfilePCREventInstrData
+}
+
+func (d *savedPCRProtectionProfileInstrData) Select(selector reflect.Value) interface{} {
```
Comment (pedronis) [inline tpm2/pcr_profile.go:781]: this should have a comment like:

// Select implements the mu.Union interface.

**Review by pedronis — APPROVED** [review]
lgtm, small comments

**Pull request merged**